src/Compilers/CSharp/Portable/Binder/Binder_Patterns.cs

```diff
@@ -155,10 +155,19 @@ private BoundPattern BindDiscardPattern(DiscardPatternSyntax node, TypeSymbol in
             BoundExpression convertedExpression = ConvertPatternExpression(
                 inputType, patternExpression, expression, out constantValueOpt, hasErrors, diagnostics);
             wasExpression = expression.Type?.IsErrorType() != true;
-            if (!convertedExpression.HasErrors && constantValueOpt == null)
+            if (!convertedExpression.HasErrors)
             {
-                diagnostics.Add(ErrorCode.ERR_ConstantExpected, patternExpression.Location);
-                hasErrors = true;
+                if (constantValueOpt == null)
+                {
+                    diagnostics.Add(ErrorCode.ERR_ConstantExpected, patternExpression.Location);
+                    hasErrors = true;
+                }
+                else if (inputType.IsPointerType() == true && Compilation.LanguageVersion < MessageID.IDS_FeatureRecursivePatterns.RequiredVersion())
+                {
+                    // before C# 8 we did not permit `pointer is null`
+                    diagnostics.Add(ErrorCode.ERR_PointerTypeInPatternMatching, patternExpression.Location);
```

Comment:
>ERR_PointerTypeInPatternMatching [](start = 46, length = 32)

nit: Would it be better to use `ERR_FeatureNotAvailableInVersion8` with an existing feature name (`IDS_FeatureRecursivePatterns` maybe)? Either way is a bit confusing to users, but we would at least hint towards language version being a fix.